Вариант решения

Nick007J opened the pull request.

Nick007J committed:
Implemented parsing

Nick007J committed:
Added parsing of 64-bit exes

Nick007J committed:
Added ImageBase to entry point address

Nick007J committed:
Removed debug stuff

Nick007J committed:
Output fix

Wunder9l commented:
Привет! Очень хорошо! Однако, если у тебя адрес точки входа указывает сразу на начало секции, то программа выведет "Section of entry point not found", что не является верным ответом.

Nick007J committed:
Error fix

Wunder9l commented:
Доброе утро! Прошу прощения за долгий ответ, я могу проверять только на рабочем компьютере (и как следствие только в будни). Программа уже выглядит практически готовой, что не может не радовать. При тестировании на файле: https://github.com/corkami/pocs/blob/master/PE/bin/96workingsections.exe была найдена ошибка в коде. Сможешь ее найти сам? Как признак - обрати внимание на некорректно отображающееся название секции.
Также есть несколько рекомендаций по стилю кода (хотя у тебя он достаточно хорошо написан):
1) Лучше исключить использование констант в коде программы (например, неплохо было бы заменить 0x10B, 0x107, 0x20B в функции GetInfoFromNTHeader на соответствующие определения из WinNt.h - IMAGE_NT_OPTIONAL_HDR32_MAGIC, IMAGE_ROM_OPTIONAL_HDR_MAGIC, IMAGE_NT_OPTIONAL_HDR64_MAGIC)
2) Если ты обратишь внимание в моем коде в конструкциях if ( ... == ... ) на первом месте стоит константа, на втором переменная. Это поможет избежать опечаток типа if ( ... = ... )  (присваивание вместо проверки на равенство).
3) Лучше все таки избежать именований типа ib, ep, а использовать полную форму imageBase, entryPoint.
4)* Все строковые константы, такие как "Not a valid DOS header\n" и т.п. в идеале выносятся в отдельный заголовочный файл. При увеличении объема кода, это позволяет быстро найти и отредактировать все вхождения конкретного сообщения, так как все сообщения собраны в одном месте.

Nick007J committed:
Buffer size increase, style fixes

На указанном файле проблема была в маленьком буфере. По-хорошему, конечно, надо было бы переделать чтение из файла с учетом значений в заголовке, но это требует исправлений в шаблоне. По примерно тем же причинам не стал вносить изменения по пункту 4*. Пункты 1, 2 и 3 переделал.

Wunder9l commented:
Да, все верно, ошибка была в том, что последний заголовок секций не был прочитан в выделенный буфер. Однако твое решение исправляет ошибку только в некоторых файлах. Что будет с твоим кодом, если скажем секций в файле не 96, а 996? А какое максимальное количество секций может быть в файле?
Ну или другой случай: вдруг у меня файл всего размером 500 байтов (к примеру). Но я захотел и ручками поменял в шестнадцатеричном редакторе значение поля file_header->NumberOfSections и поставил его значение равное скажем 99. Это приведет к краху твоей программы. Соответственно, чтобы избежать такого поведения, необходимо проверять размер выделенного буфера, именно для этого в функцию ParseFile передается его размер.
Что касается исправлений стиля. У нас на работе принят корпоративный стандарт использования (по крайней мере для c/c++) camel style. В частности, все переменные должны начинаться с маленькой буквы, и иметь вид типа imageSection, dataBufferSize (каждое слово с заглавной буквы). Все функции также имеют camel style, Однако первая буква  - заглавная (например, функция ReadFileToBuffer). Я не принуждаю тебя следовать этому стандарту, однако, важно чтобы у тебя был стандарт, которому ты следовал. Я заметил в твоем коде некую разностилицу: в одном месте ты называешь переменные p_imagebase и p_entrypoint (указывая, что это указатели), в другом месте ты уже не указываешь, что переменная - это указатель IMAGE_SECTION_HEADER* section. Наиболее показательно выглядит эта строчка:
IMAGE_OPTIONAL_HEADER32* opth32 = (IMAGE_OPTIONAL_HEADER32*)p_optheader;
Постарайся придерживаться одних правил. Будь это camelStyle или under_score, или с использованием венгерской нотации (https://ru.wikipedia.org/wiki/%D0%92%D0%B5%D0%BD%D0%B3%D0%B5%D1%80%D1%81%D0%BA%D0%B0%D1%8F_%D0%BD%D0%BE%D1%82%D0%B0%D1%86%D0%B8%D1%8F). Быть может это сейчас тебе покажется придирками, однако при расширении проекта это здорово помогает читаемости кода.

Nick007J committed:
Error fix, style fix

Переделал чтение из файла для универсальной работы. Что касается стиля: вообще-то я не сторонник венгерской нотации, в изначальном решении префиксом наделил только первый указатель функции GetInfoFromNTHeader из-за того, что это вообще нетипизированный указатель, но зачем-то вчера дописал его и другим аргументам, вообще-то я так не пишу. По поводу стиля under_score/camelCase: я использую первый в программах со стилем чистого Си, который использовал при решении (явное приведение типов, malloc вместо new и так далее).

Wunder9l commented:
Я вообще предполагал, что ты проверишь выход за размер выделенного буфера и выведешь соответствующее сообщение. Не требовалось дополнительного чтения из файла. Однако ты захотел несколько усложнить себе задачу, и набил несколько шишек.

1. Есть такое негласное правило, если делаешь однотипные действия ( более пары строк) два и более раз - повод вынести в отдельную функцию. У тебя собственно прям напрашивается написание функции типа 
bool CheckBufferSizeAndReallocate (unsigned char requieredSize, unsigned char currentBufferSize, char** buffer, bool* needToFree);
Кроме этого, я бы на твоем месте сделал функцию типа 
char* GetSectionName (char* sectionStart);

2. При работе с памятью надо тщательно отслеживать ее выделение/освобождение. Например, у тебя не освобождается выделенная память после первого if, в случае если у тебя не найдена сигнатура IMAGE_NT_SIGNATURE.

3. Вообще не очень рекомендуется иметь множественные выходы из функции (множество return). 

4. Использование goto крайне не рекомендуется (https://ru.wikipedia.org/wiki/Goto#.D0.9A.D1.80.D0.B8.D1.82.D0.B8.D0.BA.D0.B0). Если в программе требуется использование goto, то скорее всего в архитектуре приложения была допущена ошибка. Всегда можно обойтись без него. В твоем случае можно было бы использовать break + переменную типа bool

5. Завтра постараюсь найти файл, на котором твоя программа отработает неверно. Есть у меня подозрение что такой файл есть)
На самом деле, я очень доволен, что ты подошел так ответственно к задаче и решил сделать ее на совесть. Молодец!

Nick007J committed:
Style fix

С пунктом 1 соглашусь. По поводу 2, 3 и 4. Прямо в указанной статье о goto указаны случаи допустимого использования этого оператора. Большинство руководств по стилям программирования указывают те же два случая, что и википедия. Его можно было бы избежать, если отказаться от return'ов, но тогда внешний вид программы пострадает из-за ненужного нагромождения вложенных if'ов либо множества булевых переменных. Ну и мне лично нравится больше именно вариант с goto (но надо заметить, что это за счет того, что я пишу в стиле чистого Си, где нет обработки исключений и механизма деструкторов). Проблема 2 соответственно решилась за счет заменой return'ов на goto. По поводу пункта 5 - вполне допускаю и даже почти не сомневаюсь, что я где-нибудь ошибся, так что жду :)

Wunder9l commented:
Не знаю, намеренно или случайно ты использовал макрос IMAGE_FIRST_SECTION, но он уберег от одной ошибки. Дело в том, что в файле может быть меньше стандартных 16 директорий (DATA_DIRECTORIES), об этом на лекции говорил Павел Васильевич. И если бы ты получал адрес первой секции путем сложения 
( unsigned int ) nt_header + sizeof( IMAGE_NT_HEADERS ) 
то мог бы ошибиться. Правильный подход заключается в прибавлении к адресу NT-заголовка размера FileHeader'а и значения поля FileHeader.SizeOfOptionalHeader, что собственно и делает макрос IMAGE_FIRST_SECTION.
Файлов, в которых число директорий меньше стандартного, крайне мало. Но они есть, вот, например, файл вообще без директорий:
https://github.com/corkami/pocs/blob/master/PE/bin/no_dd.exe

В общем, я полностью принимаю твое решение, и поздравляю тебя с успешным выполнением задания! Закрываю твой pull-request.
